### PR TITLE
GH-35274: [Java][CI] Enable GCS on MacOS

### LINKED ISSUE #202: [Java][Dataset][MacOS] Enable GCS
### Describe the enhancement requested

Enables GCS on MacOS when building the Arrow Dataset for Java.

### Component(s)

Continuous Integration, Java

### PR DESCRIPTION
**Pull request by davisusanibar**
### Rationale for this change

Enables GCS on MacOS when building the Arrow Dataset for Java.

### What changes are included in this PR?

CI build to enable GCS

### Are these changes tested?

````java
import org.apache.arrow.dataset.file.FileFormat;
import org.apache.arrow.dataset.file.FileSystemDatasetFactory;
import org.apache.arrow.dataset.jni.NativeMemoryPool;
import org.apache.arrow.dataset.scanner.ScanOptions;
import org.apache.arrow.dataset.scanner.Scanner;
import org.apache.arrow.dataset.source.Dataset;
import org.apache.arrow.dataset.source.DatasetFactory;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.RootAllocator;
import org.apache.arrow.vector.ipc.ArrowReader;
import org.apache.arrow.vector.types.pojo.Schema;

public class Test {
  public static void main(String[] args) {
    String uri = "gs://anonymous@voltrondata-labs-datasets/nyc-taxi-tiny/year=2022/month=2/part-0.parquet"; // Google Cloud Storage
    ScanOptions options = new ScanOptions(/*batchSize*/ 32768);
    try (
        BufferAllocator allocator = new RootAllocator();
        DatasetFactory datasetFactory = new FileSystemDatasetFactory(allocator, NativeMemoryPool.getDefault(), FileFormat.PARQUET, uri);
        Dataset dataset = datasetFactory.finish();
        Scanner scanner = dataset.newScan(options);
        ArrowReader reader = scanner.scanBatches()
    ) {
      Schema schema = scanner.schema();
      System.out.println(schema); // Schema<vendor_name: Utf8, pickup_datetime: Timestamp(MILLISECOND, null), dropoff_datetime: Timestamp(MILLISECOND, null), passenger_count: Int(64, true), trip_distance: FloatingPoint(DOUBLE), pickup_longitude: FloatingPoint(DOUBLE), pickup_latitude: FloatingPoint(DOUBLE), rate_code: Utf8, store_and_fwd: Utf8, dropoff_longitude: FloatingPoint(DOUBLE), dropoff_latitude: FloatingPoint(DOUBLE), payment_type: Utf8, fare_amount: FloatingPoint(DOUBLE), extra: FloatingPoint(DOUBLE), mta_tax: FloatingPoint(DOUBLE), tip_amount: FloatingPoint(DOUBLE), tolls_amount: FloatingPoint(DOUBLE), total_amount: FloatingPoint(DOUBLE), improvement_surcharge: FloatingPoint(DOUBLE), congestion_surcharge: FloatingPoint(DOUBLE), pickup_location_id: Int(64, true), dropoff_location_id: Int(64, true)>
      while (reader.loadNextBatch()) {
        System.out.println("RowCount: " + reader.getVectorSchemaRoot().getRowCount()); //2979
      }
    } catch (Exception e) {
      e.printStackTrace();
    }
  }
}
````

### Are there any user-facing changes?

No

* Closes: apache/arrow-java#202
* GitHub Issue: #202

### CODE REVIEW

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

File: cpp/build-support/run-test.sh [hunk 1]
```diff
@@ -60,6 +60,9 @@ rm -f $LOGFILE $LOGFILE.gz
 
 pipe_cmd=cat
 
+# Install the gcs testbench
+sh ../../../../../ci/scripts/install_gcs_testbench.sh default
```
Comment (assignUser) [inline cpp/build-support/run-test.sh:64]: I think this is quite confusing without following the previous code in detail. I think doing this in relation to `$ROOT` would make it a bit clearer?
```suggestion
sh $ROOT/../ci/scripts/install_gcs_testbench.sh default
```

**Review by assignUser — COMMENTED** [review]
A minor change

**Comment by davisusanibar** [comment]
I am seeing error on MacOS aarch_64 with message `dyld: lazy symbol binding failed: Symbol not found: _curl_multi_poll` @kou / @assignUser  Do you have any idea about the reason of?

```c++
Running arrow-gcsfs-test, redirecting output into /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/arrow/cpp-build/cpp/build/test-logs/arrow-gcsfs-test.txt (attempt 1/1)
/Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/arrow/cpp/build-support/run-test.sh: line 91: 33793 Abort trap: 6           $TEST_EXECUTABLE "$@" > $LOGFILE.raw 2>&1
INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://localhost:63593
INFO:werkzeug:Press CTRL+C to quit
INFO:werkzeug: * Restarting with stat
Running main() from /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/arrow/cpp-build/cpp/googletest_ep-prefix/src/googletest_ep/googletest/src/gtest_main.cc
[==========] Running 99 tests from 3 test suites.
[----------] Global test environment set-up.
[----------] 26 tests from TestGCSFSGeneric
[ RUN      ] TestGCSFSGeneric.Empty
dyld: lazy symbol binding failed: Symbol not found: _curl_multi_poll
  Referenced from: /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/arrow/cpp-build/cpp/release/arrow-gcsfs-test
  Expected in: /usr/lib/libcurl.4.dylib

dyld: Symbol not found: _curl_multi_poll
  Referenced from: /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/arrow/cpp-build/cpp/release/arrow-gcsfs-test
  Expected in: /usr/lib/libcurl.4.dylib

...
...

Total Test time (real) = 121.27 sec

The following tests FAILED:
	 61 - arrow-gcsfs-test (Failed)
Errors while running CTest
```

**Comment by assignUser** [comment]
I think the installed curl version is too old. It looks like libcurl 4 is installed but multi pool was introduced in 7 https://curl.se/libcurl/c/curl_multi_poll.html

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java-jars

File: cpp/thirdparty/versions.txt [hunk 2]
```diff
@@ -23,8 +23,8 @@
 # `DEPENDENCIES` array (see the comment on top of the declaration for the
 # format).
 
-ARROW_ABSL_BUILD_VERSION=20211102.0
-ARROW_ABSL_BUILD_SHA256_CHECKSUM=dcf71b9cba8dc0ca9940c4b316a0c796be8fab42b070bb6b7cab62b48f0e66c4
+ARROW_ABSL_BUILD_VERSION=20230125.2
```
Comment (lidavidm) [inline cpp/thirdparty/versions.txt:26]: I'm not sure we can just update this without also possibly updating gRPC/Google Cloud SDK

Comment (lidavidm) [reply cpp/thirdparty/versions.txt:26]: Hmm, seems Abseil promises compatibility now. So it looks like we're safe here, at least it compiles.

**Comment by assignUser** [comment]
This should fix the m1 issues
```
--- a/dev/tasks/java-jars/github.yml
+++ b/dev/tasks/java-jars/github.yml
@@ -108,6 +108,7 @@ jobs:
           set -e
           # make brew Java available to CMake
           if [ "{{ arch }}" = "aarch_64" ]; then
+            export CURL_ROOT=$(brew --prefix curl)
             export JAVA_HOME=$(brew --prefix openjdk@11)/libexec/openjdk.jdk/Contents/Home
           fi
           arrow/ci/scripts/java_jni_macos_build.sh \
```

**Comment by assignUser** [comment]
@github-actions crossbow submit java-jars

**Comment by davisusanibar** [comment]
> Revision: [26d808d](https://github.com/apache/arrow/commit/26d808d39314845889b179d53b90367d1ae4d119)
> 
> Submitted crossbow builds: [ursacomputing/crossbow @ actions-ef819b2719](https://github.com/ursacomputing/crossbow/branches/all?query=actions-ef819b2719)
> 
> Task	Status
> java-jars	[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/4961934231/jobs/8879348028)

Thank you @assignUser !

**Review by assignUser — APPROVED** [review]
imo this is merge ready now +1

File: java/Brewfile [hunk 3]
```diff
@@ -17,3 +17,4 @@
 
 brew "openjdk@11"
 brew "sccache"
+brew "curl"
```
Comment (pitrou) [inline java/Brewfile:20]: Perhaps keep this file in alphabetical order?

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java-jars

File: java/Brewfile (hunk 3 above)
Comment (davisusanibar) [reply java/Brewfile]: Juste deleted, that was added for testing purpose that at the end we add CURL_ROOT on the github.yml file

File: cpp/build-support/run-test.sh [hunk 4]
```diff
@@ -60,6 +60,9 @@ rm -f $LOGFILE $LOGFILE.gz
 
 pipe_cmd=cat
 
+# Install the gcs testbench
+sh $ROOT/../ci/scripts/install_gcs_testbench.sh default
```
[the remark this answers is not in the document]
Comment (davisusanibar) [reply cpp/build-support/run-test.sh]: If we enable ARROW_GCS some test need some software base that is installed by this shell. Just moved inside of MacOS scope.

File: ci/scripts/java_jni_macos_build.sh [hunk 5]
```diff
@@ -75,15 +77,18 @@ cmake \
   -DARROW_DEPENDENCY_USE_SHARED=OFF \
   -DARROW_GANDIVA=${ARROW_GANDIVA} \
   -DARROW_GANDIVA_STATIC_LIBSTDCPP=ON \
+  -DARROW_GCS=${ARROW_GCS} \
   -DARROW_ORC=${ARROW_ORC} \
   -DARROW_PARQUET=${ARROW_PARQUET} \
   -DARROW_S3=${ARROW_S3} \
   -DARROW_USE_CCACHE=${ARROW_USE_CCACHE} \
+  -DARROW_WITH_SNAPPY=ON \
```
[the remark this answers is not in the document]
Comment (davisusanibar) [reply ci/scripts/java_jni_macos_build.sh:86]: At Arrow build, ARROW_WITH_SNAPPY is not added by default and GCS use that codec by default.

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java-jars

**Comment by davisusanibar** [comment]
@lidavidm do you have some insight about what I could review to solve this error https://github.com/apache/arrow/actions/runs/5274483912/jobs/9538989549?pr=35277:

```
[70/122] Generating arrow-flight-glib/ArrowFlight-1.0.gir with a custom command (wrapped by meson to set env)
FAILED: arrow-flight-glib/ArrowFlight-1.0.gir 
env PKG_CONFIG_PATH=/usr/local/lib/pkgconfig:/build/c_glib/meson-uninstalled /usr/bin/g-ir-scanner --quiet --no-libtool --namespace=ArrowFlight --nsversion=1.0 --warn-all --output arrow-flight-glib/ArrowFlight-1.0.gir --c-include=arrow-flight-glib/arrow-flight-glib.h --warn-all --include-uninstalled=./arrow-glib/Arrow-1.0.gir -I/arrow/c_glib/arrow-flight-glib -I/build/c_glib/arrow-flight-glib -I/arrow/c_glib/. -I/build/c_glib/. --filelist=/build/c_glib/arrow-flight-glib/libarrow-flight-glib.so.1300.0.0.p/ArrowFlight_1.0_gir_filelist --include=Arrow-1.0 --symbol-prefix=gaflight --identifier-prefix=GAFlight --pkg-export=arrow-flight-glib --cflags-begin -I/arrow/c_glib/. -I/build/c_glib/. -I/usr/local/include -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/gobject-introspection-1.0 -DARROW_NO_DEPRECATED_API --cflags-end --add-include-path=/build/c_glib/arrow-glib --add-include-path=/usr/share/gir-1.0 -L/build/c_glib/arrow-flight-glib --library arrow-flight-glib -L/build/c_glib/arrow-glib -L/usr/local/lib -L/usr/local/lib --extra-library=arrow_flight --extra-library=arrow --extra-library=arrow_acero --extra-library=gobject-2.0 --extra-library=glib-2.0 --extra-library=girepository-1.0 --sources-top-dirs /arrow/c_glib/ --sources-top-dirs /build/c_glib/ --warn-error
/usr/bin/ld: /usr/local/lib/libarrow_flight.so: undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::CrcCordState()'
/usr/bin/ld: /usr/local/lib/libarrow_flight.so: undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::operator=(absl::lts_20230125::crc_internal::CrcCordState&&)'
/usr/bin/ld: /usr/local/lib/libarrow_flight.so: undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::CrcCordState(absl::lts_20230125::crc_internal::CrcCordState&&)'
/usr/bin/ld: /usr/local/lib/libarrow_flight.so: undefined reference to `absl::lts_20230125::base_internal::StrError[abi:cxx11](int)'
/usr/bin/ld: /usr/local/lib/libarrow_flight.so: undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::~CrcCordState()'
/usr/bin/ld: /usr/local/lib/libarrow_flight.so: undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::Checksum() const'
collect2: error: ld returned 1 exit status
linking of temporary binary failed: Command '['x86_64-linux-gnu-gcc', '-pthread', '-o', '/build/c_glib/tmp-introspectci4byn1o/ArrowFlight-1.0', '-DARROW_NO_DEPRECATED_API', '/build/c_glib/tmp-introspectci4byn1o/ArrowFlight-1.0.o', '-L.', '-Wl,-rpath,.', '-Wl,--no-as-needed', '-L/build/c_glib/arrow-flight-glib', '-Wl,-rpath,/build/c_glib/arrow-flight-glib', '-L/build/c_glib/arrow-glib', '-Wl,-rpath,/build/c_glib/arrow-glib', '-L/usr/local/lib', '-Wl,-rpath,/usr/local/lib', '-L/usr/local/lib', '-Wl,-rpath,/usr/local/lib', '-larrow-flight-glib', '-larrow_flight', '-larrow', '-larrow_acero', '-lgobject-2.0', '-lglib-2.0', '-lgirepository-1.0', '-lgio-2.0', '-lgobject-2.0', '-Wl,--export-dynamic', '-lgmodule-2.0', '-pthread', '-lglib-2.0']' returned non-zero exit status 1.
[71/122] Compiling C++ object gandiva-glib/libgandiva-glib.so.1300.0.0.p/expression.cpp.o
[72/122] Compiling C++ object arrow-flight-sql-glib/libarrow-flight-sql-glib.so.1300.0.0.p/server.cpp.o
ninja: build stopped: subcommand failed.
```

**Comment by kou** [comment]
Hmm. It's not reproduced on local...

It seems that the error message show that we need to add Abseil's CRC related modules to https://github.com/apache/arrow/blob/main/cpp/cmake_modules/ThirdpartyToolchain.cmake#L2856-L3756 .

BTW, can we use 20230125.3 https://github.com/abseil/abseil-cpp/releases/tag/20230125.3 ?

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java-jars

**Comment by davisusanibar** [comment]
@lidavidm please could you help me to move this PR to draft?, there are some problems on the CI that I need to test and probably need to call it again and again. Thank you in advance

**Comment by kou** [comment]
You can find the "Convert to draft" link in the right side bar.

**Comment by davisusanibar** [comment]
> You can find the "Convert to draft" link in the right side bar.

Thank you

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java-jars

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java-jars

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java-jars

**Comment by davisusanibar** [comment]
In order to define a better alternative solution, the current pull request has been closed.

**Pull request closed without merging**